### PR TITLE
MINE

### PR DESCRIPTION
**Pull request by AnejSvete**
First draft of MINE together with the compatible data classes for dSprites. 
I think the implementation is correct, but some things should probably be restructured.
Largely inspired by the great implementation here: https://github.com/gtegner/mine-pytorch

### CODE REVIEW

File: latte/dataset/dsprites.py [hunk 1]
```diff
@@ -1,19 +1,50 @@
...
+
+factor_names = {
+    DSpritesFactor.SHAPE: "Shape",
+    DSpritesFactor.SCALE: "Scale",
+    DSpritesFactor.ORIENTATION: "Orientation",
+    DSpritesFactor.X: "Position x",
+    DSpritesFactor.Y: "Position y",
+}
+
+
```
Comment (pawel-czyz) [inline latte/dataset/dsprites.py:38]: I now know what it does, but it would be good to add a one-line comment :slightly_smiling_face: 
(As we're likely to forget it in 3-month time and checking the code requires 2 minutes instead of 2 seconds :wink: )

File: latte/dataset/dsprites.py [hunk 2]
```diff
@@ -125,6 +164,132 @@ def load_dsprites(filepath: Union[str, pathlib.Path] = DEFAULT_TARGET, download_
...
+    assert method in [
+        "PCA",
+        "FA",
+        "FactorAnalysis",
+    ], f"Method {method} is not supported, only PCA and FactorAnalysis supported."
+
+    if factors is None:
+        # By default, return all factors
+        factors = [
```
Comment (pawel-czyz) [inline latte/dataset/dsprites.py:240]: I think we can do this in a simpler way – what do you think about `list(factor_names.keys())`?

File: latte/dataset/dsprites.py [hunk 3]
```diff
@@ -125,6 +164,132 @@ def load_dsprites(filepath: Union[str, pathlib.Path] = DEFAULT_TARGET, download_
...
+            DSpritesFactor.X,
+            DSpritesFactor.Y,
+        ]
+
+    # Load raw data
+    data = load_dsprites(filepath, download_ok)
+
+    # Flatten and filter the data
+    X, y, factor2idx = prepare_data(
+        data, factors=factors, ignored_factors=ignored_factors if ignored_factors is not None else []
```
Comment (pawel-czyz) [inline latte/dataset/dsprites.py:253]: Consider making `ignored_factors` an `Optional[List[DSpritesFactor]` in `prepare_data`, so that this `if` won't be needed in this function.

File: latte/models/mine.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+EPSILON = 1e-6
+
+
+class MINELossType(Enum):
+    MINE = 1
+    F_DIV = 2
+    MINE_BIASED = 3
+
+
+class MINEBiasCorrection(torch.autograd.Function):
```
Comment (pawel-czyz) [inline latte/models/mine.py:26]: This is gold :+1:

File: latte/models/mine.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    def forward(self, *x: torch.Tensor) -> torch.Tensor:
+        for module in self._modules.values():
+            if isinstance(x, tuple):
+                x = module(*x)
+            else:
+                x = module(x)
+        return x
+
+
+class StatisticsNetwork(nn.Module):
```
Comment (pawel-czyz) [inline latte/models/mine.py:92]: Wonderful!

**Review by pawel-czyz — APPROVED** [review]
Hi Anej!

The whole PR is just wonderful! 

Addressing your questions:
1. Tests: 
2. The structure: right now it looks good to me. We'll probably revisit the structure later, but some amount of restructuring is always inevitable – we'll learn how to adjust it when we have the core ready (another dataset or two, some latent spaces, and plugging everything into the optimization loop, preferably over Stiefel manifol/Grassmanian. We'll talk about it more on Monday).
3. This form of data splitting is more than sufficient at this stage, I think. We'll adapt it if needed :slightly_smiling_face: (And when we have more datasets, it'll be easier to know how to structure this code).

File: latte/models/mine.py (hunk 4 above)
Comment (AnejSvete) [reply latte/models/mine.py:28]: 😃

File: latte/dataset/dsprites.py [hunk 6]
```diff
@@ -125,6 +164,132 @@ def load_dsprites(filepath: Union[str, pathlib.Path] = DEFAULT_TARGET, download_
...
+    )
+
+    # Reduce the dimensionality
+    if method == "PCA":
```
[the remark this answers is not in the document]
Comment (AnejSvete) [reply latte/dataset/dsprites.py:259]: I think the second option is more generalizable for later, I'll do that :)

File: latte/dataset/dsprites.py (hunk 3 above)
Comment (AnejSvete) [reply latte/dataset/dsprites.py]: Did that. I'm not sure if this can still be optimised, but now I check if `ignored_factors is None` in `prepare_data`. Is that okay?

File: latte/dataset/dsprites.py [hunk 7]
```diff
@@ -125,6 +164,132 @@ def load_dsprites(filepath: Union[str, pathlib.Path] = DEFAULT_TARGET, download_
...
+    X = StandardScaler().fit_transform(X)
+    X = trans.fit_transform(X)
+
+    return PreprocessedDSpritesDataset(
+        X, y, factor2idx, pca_ratios=trans.explained_variance_ratio_ if method == "PCA" else None
+    )
+
+
+# TODO (Anej)
+def split(
```
[the remark this answers is not in the document]
Comment (AnejSvete) [reply latte/dataset/dsprites.py:273]: Thanks for the tip! Implemented the generator :)

File: latte/dataset/dsprites.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AnejSvete) [reply latte/dataset/dsprites.py:259]: Hmm, on a second thought, I think this would require a bit too much work upstream (also for later).
I'll just implement a function for it.

File: latte/dataset/dsprites.py (hunk 3 above)
Comment (pawel-czyz) [reply latte/dataset/dsprites.py]: Good idea! :+1:

File: latte/dataset/dsprites.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pawel-czyz) [reply latte/dataset/dsprites.py]: Nice!

File: latte/models/callbacks.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    """A callback for keeping track of the weights throughout training.
+    Used for unit testing."""
+
+    def __init__(self):
+        self.epoch_weights = defaultdict(list)
+
+    def on_validation_epoch_end(self, trainer: pl.Trainer, module: mine.MINE) -> None:
+        for ii, layer in enumerate(module.T.S):
+            if hasattr(layer, "weight"):
+                self.epoch_weights[ii].append(copy.deepcopy(layer.weight))
```
Comment (pawel-czyz) [inline latte/models/callbacks.py:29]: Hmmm, `layer.weight` is a tensor and I'm not sure how deepcopy interacts with it. In particular,

1. It may be on GPU, so I expect that deepcopy may not handle that.
2. Can it have `require_gradients=True`? If so, it'll be connected to the computation graph and we don't need it.

Perhaps it'd be better to call something like `cpu().detach().numpy().copy()` on it.

File: latte/models/datamodules.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    """A generic DataModule for serving data from two distributions to MINE."""
+
+    def __init__(
+        self,
+        X: np.ndarray,
+        Z: np.ndarray,
+        p_train: float = 0.7,
+        p_val: float = 0.1,
+        batch_size: int = 32,
+        test_batch_size=512,
```
Comment (pawel-czyz) [inline latte/models/datamodules.py:24]: nit: missing type annotation

File: latte/models/mine.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    running_mean = exp_t_marginal if running_mean == 0 else (1 - alpha) * running_mean + alpha * exp_t_marginal
+
+    y = MINEBiasCorrection.apply(t_marginal, running_mean)
+
+    return torch.mean(t) - y, running_mean
+
+
+def f_div_objective(t: torch.Tensor, t_marginal: torch.Tensor) -> torch.Tensor:
+    """The f-divergence upper-bound-based loss."""
+    return torch.mean(t) - torch.mean(torch.exp(t_marginal - 1))
```
Comment (pawel-czyz) [inline latte/models/mine.py:65]: How numerically stable is this? Shall we try to move `t_marginal.max()` before the mean and subtract it from `t_marginal`?
(I'm unsure how the original implementation looks like, perhaps this works very well, so just asking :slightly_smiling_face: )

**Review by pawel-czyz — APPROVED** [review]
This is wonderful, thank you a lot!

I had some minor comments, the only concerning one regards the weight logging (whether they are disconnected from the computational graph).

File: latte/models/mine.py (hunk 10 above)
Comment (AnejSvete) [reply latte/models/mine.py:65]: Hmm, so the implementation here https://github.com/gtegner/mine-pytorch/blob/master/mine/models/mine.py is the same as I have. It's not the original authors but it's a very faithful implementation.
I don't think it's very important since I don't think we'll be using this method much any way, but did you mean replacing the second term with `torch.mean(torch.exp(t_marginal - torch.max(t_marginal) - 1)) * torch.max(t_marginal)`?

File: latte/models/callbacks.py (hunk 8 above)
Comment (AnejSvete) [reply latte/models/callbacks.py:29]: Good point, fixed!

File: latte/models/mine.py (hunk 10 above)
Comment (pawel-czyz) [reply latte/models/mine.py:65]: Yes, this is what I meant. On the second thought, I'm not sure if it'd be more stable. As you know that it works, it's looks good for me as well – please, merge. Good job! :slightly_smiling_face:
